README.md
# TypedStructLegacyReflection

[![Build Status](https://travis-ci.com/ejpcmac/typed_struct_legacy_reflection.svg?branch=develop)](https://travis-ci.com/ejpcmac/typed_struct_legacy_reflection)
[![hex.pm version](http://img.shields.io/hexpm/v/typed_struct_legacy_reflection.svg?style=flat)](https://hex.pm/packages/typed_struct_legacy_reflection)

TypedStructLegacyReflection is a
[TypedStruct](https://github.com/ejpcmac/typed_struct) plugin defining the
reflection functions that used to be built-in before TypedStruct 0.2.0.

## Setup

To use this plugin in your project, add this to your Mix dependencies:

```elixir
{:typed_struct_legacy_reflection, "~> 1.0.0"}
```

If you do not plan to compile modules using this TypedStruct plugin at runtime,
you can add `runtime: false` to the dependency tuple as it is only used during
compilation.

## Usage

To use this plugin in a typed struct, simply register it in the `typedstruct`
block:

```elixir
defmodule StructWithReflection do
  use TypedStruct

  typedstruct do
    plugin TypedStructLegacyReflection

    field :a_field, term()
    field :another_field, String.t()
  end
end
```

Three functions are then defined in your typed struct, like TypedStruct did in
v0.1.x:

* `__keys__/0` - returns the keys of the struct
* `__defaults__/0` - returns the default value for each field
* `__types__/0` - returns the quoted type for each field

For instance:

```elixir
iex(1)> defmodule Demo do
...(1)>   use TypedStruct
...(1)>
...(1)>   typedstruct do
...(1)>     plugin TypedStructLegacyReflection
...(1)>
...(1)>     field :a_field, String.t()
...(1)>     field :with_default, integer(), default: 7
...(1)>   end
...(1)> end
{:module, Demo,
 <<70, 79, 82, 49, 0, 0, 8, 184, 66, 69, 65, 77, 65, 116, 85, 56, 0, 0, 0, 241,
   0, 0, 0, 24, 11, 69, 108, 105, 120, 105, 114, 46, 68, 101, 109, 111, 8, 95,
   95, 105, 110, 102, 111, 95, 95, 7, 99, ...>>,
 [{TypedStructLegacyReflection, []}]}
iex(2)> Demo.__keys__()
[:a_field, :with_default]
iex(3)> Demo.__defaults__()
[a_field: nil, with_default: 7]
iex(4)> Demo.__types__()
[
  a_field: {:|, [],
   [
     {{:., [line: 7],
       [{:__aliases__, [counter: {Demo, 3}, line: 7], [:String]}, :t]},
      [line: 7], []},
     nil
   ]},
  with_default: {:integer, [line: 8], []}
]
```

## [Contributing](CONTRIBUTING.md)

Before contributing to this project, please read the
[CONTRIBUTING.md](CONTRIBUTING.md).

## License

Copyright © 2018, 2020 Jean-Philippe Cugnet

This project is licensed under the [MIT license](LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic
Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2020-05-31

### Added

* Implement the reflection API from TypedStruct 0.1.x.

[1.0.0]: https://github.com/ejpcmac/typed_struct_legacy_reflection/releases/tag/v1.0.0


test/typed_struct_legacy_reflection_test.exs
defmodule TypedStructLegacyReflectionTest do
  use ExUnit.Case

  ############################################################################
  ##                               Test data                                ##
  ############################################################################

  defmodule TestStruct do
    use TypedStruct

    typedstruct do
      plugin TypedStructLegacyReflection

      field :int, integer()
      field :string, String.t()
      field :string_with_default, String.t(), default: "default"
      field :mandatory_int, integer(), enforce: true
    end
  end

  ############################################################################
  ##                             Standard cases                             ##
  ############################################################################

  test "generates a function to get the struct keys" do
    assert TestStruct.__keys__() == [
             :int,
             :string,
             :string_with_default,
             :mandatory_int
           ]
  end

  test "generates a function to get the struct defaults" do
    assert TestStruct.__defaults__() == [
             int: nil,
             string: nil,
             string_with_default: "default",
             mandatory_int: nil
           ]
  end

  test "generates a function to get the struct types" do
    types =
      quote do
        [
          int: integer() | nil,
          string: String.t() | nil,
          string_with_default: String.t(),
          mandatory_int: integer()
        ]
      end

    assert delete_context(TestStruct.__types__()) == delete_context(types)
  end

  ############################################################################
  ##                                Helpers                                 ##
  ############################################################################

  # Deletes the context from a quoted expression.
  defp delete_context(list) when is_list(list),
    do: Enum.map(list, &delete_context/1)

  defp delete_context({a, b}),
    do: {delete_context(a), delete_context(b)}

  defp delete_context({fun, _context, args}),
    do: {delete_context(fun), [], delete_context(args)}

  defp delete_context(other), do: other
end


mix.exs
defmodule TypedStructLegacyReflection.MixProject do
  use Mix.Project

  @version "1.0.0"
  @repo_url "https://github.com/ejpcmac/typed_struct_legacy_reflection"

  def project do
    [
      app: :typed_struct_legacy_reflection,
      version: @version,
      elixir: "~> 1.6",
      start_permanent: Mix.env() == :prod,
      deps: deps(),

      # Tools
      dialyzer: dialyzer(),
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: cli_env(),

      # Docs
      docs: [
        main: "TypedStructLegacyReflection",
        source_url: @repo_url,
        source_ref: "v#{@version}"
      ],

      # Package
      package: package(),
      description: """
      A TypedStruct plugin re-enabling the legacy reflection functions from
      TypedStruct 0.1.x.
      """
    ]
  end

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp deps do
    [
      # Development dependencies
      {:ex_check, "~> 0.11.0", only: :dev, runtime: false},
      {:credo, "~> 1.0", only: :dev, runtime: false},
      {:dialyxir, "~> 1.0-rc", only: :dev, runtime: false},
      {:excoveralls, ">= 0.0.0", only: :test, runtime: false},
      {:mix_test_watch, ">= 0.0.0", only: :test, runtime: false},
      {:ex_unit_notifier, ">= 0.0.0", only: :test, runtime: false},

      # Project dependencies
      {:typed_struct, "~> 0.2.0"},

      # Documentation dependencies
      {:ex_doc, "~> 0.19", only: :docs, runtime: false}
    ]
  end

  # Dialyzer configuration
  defp dialyzer do
    [
      # Use a custom PLT directory for continuous integration caching.
      plt_core_path: System.get_env("PLT_DIR"),
      plt_file: plt_file(),
      plt_add_deps: :app_tree,
      flags: [
        :unmatched_returns,
        :error_handling,
        :race_conditions
      ],
      ignore_warnings: ".dialyzer_ignore"
    ]
  end

  defp plt_file do
    case System.get_env("PLT_DIR") do
      nil ->
        nil

      plt_dir ->
        {:no_warn, Path.join(plt_dir, "typed_struct_legacy_reflection.plt")}
    end
  end

  defp cli_env do
    [
      # Run mix test.watch in `:test` env.
      "test.watch": :test,

      # Always run Coveralls Mix tasks in `:test` env.
      coveralls: :test,
      "coveralls.detail": :test,
      "coveralls.html": :test,

      # Use a custom env for docs.
      docs: :docs
    ]
  end

  defp package do
    [
      licenses: ["MIT"],
      links: %{"GitHub" => @repo_url}
    ]
  end

  # Helper to add a development revision to the version. Do NOT make a call to
  # Git this way in a production release!!
  def dev do
    with {rev, 0} <-
           System.cmd("git", ["rev-parse", "--short", "HEAD"],
             stderr_to_stdout: true
           ) do
      "-dev+" <> String.trim(rev)
    else
      _ -> "-dev"
    end
  end
end


.credo.exs
# This file contains the configuration for Credo and you are probably reading
# this after creating it with `mix credo.gen.config`.
#
# If you find anything wrong or unclear in this file, please report an
# issue on GitHub: https://github.com/rrrene/credo/issues
#
%{
  #
  # You can have as many configs as you like in the `configs:` field.
  configs: [
    %{
      #
      # Run any exec using `mix credo -C <name>`. If no exec name is given
      # "default" is used.
      #
      name: "default",
      #
      # These are the files included in the analysis:
      files: %{
        #
        # You can give explicit globs or simply directories.
        # In the latter case `**/*.{ex,exs}` will be used.
        #
        included: ["lib/", "src/", "test/", "web/", "apps/"],
        excluded: [~r"/_build/", ~r"/deps/", ~r"/node_modules/"]
      },
      #
      # If you create your own checks, you must specify the source files for
      # them here, so they can be loaded by Credo before running the analysis.
      #
      requires: [],
      #
      # If you want to enforce a style guide and need a more traditional linting
      # experience, you can change `strict` to `true` below:
      #
      strict: true,
      #
      # If you want to use uncolored output by default, you can change `color`
      # to `false` below:
      #
      color: true,
      #
      # You can customize the parameters of any check by adding a second element
      # to the tuple.
      #
      # To disable a check put `false` as second element:
      #
      #     {Credo.Check.Design.DuplicatedCode, false}
      #
      checks: [
        #
        ## Consistency Checks
        #
        {Credo.Check.Consistency.ExceptionNames, []},
        {Credo.Check.Consistency.LineEndings, []},
        {Credo.Check.Consistency.ParameterPatternMatching, []},
        {Credo.Check.Consistency.SpaceAroundOperators, []},
        {Credo.Check.Consistency.SpaceInParentheses, []},
        {Credo.Check.Consistency.TabsOrSpaces, []},

        #
        ## Design Checks
        #
        # You can customize the priority of any check
        # Priority values are: `low, normal, high, higher`
        #
        {Credo.Check.Design.AliasUsage,
         [
           priority: :low,
           if_nested_deeper_than: 2,
           if_called_more_often_than: 0
         ]},
        # For some checks, you can also set other parameters
        #
        # If you don't want the `setup` and `test` macro calls in ExUnit tests
        # or the `schema` macro in Ecto schemas to trigger DuplicatedCode, just
        # set the `excluded_macros` parameter to `[:schema, :setup, :test]`.
        #
        {Credo.Check.Design.DuplicatedCode, [excluded_macros: []]},
        # You can also customize the exit_status of each check.
        # If you don't want TODO comments to cause `mix credo` to fail, just
        # set this value to 0 (zero).
        #
        {Credo.Check.Design.TagTODO, [exit_status: 0]},
        {Credo.Check.Design.TagFIXME, []},

        #
        ## Readability Checks
        #
        {Credo.Check.Readability.AliasOrder, []},
        {Credo.Check.Readability.FunctionNames, []},
        {Credo.Check.Readability.LargeNumbers, []},
        {Credo.Check.Readability.MaxLineLength,
         [priority: :low, max_length: 80]},
        {Credo.Check.Readability.ModuleAttributeNames, []},
        {Credo.Check.Readability.ModuleDoc, []},
        {Credo.Check.Readability.ModuleNames, []},
        {Credo.Check.Readability.ParenthesesInCondition, []},
        {Credo.Check.Readability.ParenthesesOnZeroArityDefs, []},
        {Credo.Check.Readability.PredicateFunctionNames, []},
        {Credo.Check.Readability.PreferImplicitTry, []},
        {Credo.Check.Readability.RedundantBlankLines, []},
        {Credo.Check.Readability.Semicolons, []},
        {Credo.Check.Readability.SpaceAfterCommas, []},
        {Credo.Check.Readability.StringSigils, []},
        {Credo.Check.Readability.TrailingBlankLine, []},
        {Credo.Check.Readability.TrailingWhiteSpace, []},
        {Credo.Check.Readability.VariableNames, []},

        #
        ## Refactoring Opportunities
        #
        {Credo.Check.Refactor.CondStatements, []},
        {Credo.Check.Refactor.CyclomaticComplexity, []},
        {Credo.Check.Refactor.FunctionArity, []},
        {Credo.Check.Refactor.LongQuoteBlocks, []},
        {Credo.Check.Refactor.MapInto, false},
        {Credo.Check.Refactor.MatchInCondition, []},
        {Credo.Check.Refactor.NegatedConditionsInUnless, []},
        {Credo.Check.Refactor.NegatedConditionsWithElse, []},
        {Credo.Check.Refactor.Nesting, []},
        {Credo.Check.Refactor.PipeChainStart,
         [
           excluded_argument_types: [:atom, :binary, :fn, :keyword],
           excluded_functions: []
         ]},
        {Credo.Check.Refactor.UnlessWithElse, []},

        #
        ## Warnings
        #
        {Credo.Check.Warning.BoolOperationOnSameValues, []},
        {Credo.Check.Warning.ExpensiveEmptyEnumCheck, []},
        {Credo.Check.Warning.IExPry, []},
        {Credo.Check.Warning.IoInspect, []},
        {Credo.Check.Warning.LazyLogging, false},
        {Credo.Check.Warning.OperationOnSameValues, []},
        {Credo.Check.Warning.OperationWithConstantResult, []},
        {Credo.Check.Warning.RaiseInsideRescue, []},
        {Credo.Check.Warning.UnusedEnumOperation, []},
        {Credo.Check.Warning.UnusedFileOperation, []},
        {Credo.Check.Warning.UnusedKeywordOperation, []},
        {Credo.Check.Warning.UnusedListOperation, []},
        {Credo.Check.Warning.UnusedPathOperation, []},
        {Credo.Check.Warning.UnusedRegexOperation, []},
        {Credo.Check.Warning.UnusedStringOperation, []},
        {Credo.Check.Warning.UnusedTupleOperation, []},

        #
        # Controversial and experimental checks (opt-in, just remove `, false`)
        #
        {Credo.Check.Consistency.MultiAliasImportRequireUse, false},
        {Credo.Check.Readability.Specs, false},
        {Credo.Check.Refactor.ABCSize, false},
        {Credo.Check.Refactor.AppendSingleItem, false},
        {Credo.Check.Refactor.DoubleBooleanNegation, false},
        {Credo.Check.Refactor.VariableRebinding, false},
        {Credo.Check.Warning.MapGetUnsafePass, false},
        {Credo.Check.Warning.UnsafeToAtom, false}

        #
        # Custom checks can be created using `mix credo.gen.check`.
        #
      ]
    }
  ]
}


lib/typed_struct_legacy_reflection.ex
defmodule TypedStructLegacyReflection do
  @moduledoc """
  TypedStructLegacyReflection is a
  [TypedStruct](https://github.com/ejpcmac/typed_struct) plugin defining the
  reflection functions that used to be built-in before TypedStruct 0.2.0.

  ## Setup

  To use this plugin in your project, add this to your Mix dependencies:

      {:typed_struct_legacy_reflection, "~> #{Mix.Project.config()[:version]}"}

  If you do not plan to compile modules using this TypedStruct plugin at
  runtime, you can add `runtime: false` to the dependency tuple as it is only
  used during compilation.

  ## Usage

  To use this plugin in a typed struct, simply register it in the `typedstruct`
  block:

      defmodule StructWithReflection do
        use TypedStruct

        typedstruct do
          plugin TypedStructLegacyReflection

          field :a_field, term()
          field :another_field, String.t()
        end
      end

  Three functions are then defined in your typed struct, like TypedStruct did in
  v0.1.x:

  * `__keys__/0` - returns the keys of the struct
  * `__defaults__/0` - returns the default value for each field
  * `__types__/0` - returns the quoted type for each field

  For instance:

      iex(1)> defmodule Demo do
      ...(1)>   use TypedStruct
      ...(1)>
      ...(1)>   typedstruct do
      ...(1)>     plugin TypedStructLegacyReflection
      ...(1)>
      ...(1)>     field :a_field, String.t()
      ...(1)>     field :with_default, integer(), default: 7
      ...(1)>   end
      ...(1)> end
      {:module, Demo,
      <<70, 79, 82, 49, 0, 0, 8, 184, 66, 69, 65, 77, 65, 116, 85, 56, 0, 0, 0, 241,
        0, 0, 0, 24, 11, 69, 108, 105, 120, 105, 114, 46, 68, 101, 109, 111, 8, 95,
        95, 105, 110, 102, 111, 95, 95, 7, 99, ...>>,
      [{TypedStructLegacyReflection, []}]}
      iex(2)> Demo.__keys__()
      [:a_field, :with_default]
      iex(3)> Demo.__defaults__()
      [a_field: nil, with_default: 7]
      iex(4)> Demo.__types__()
      [
        a_field: {:|, [],
        [
          {{:., [line: 7],
            [{:__aliases__, [counter: {Demo, 3}, line: 7], [:String]}, :t]},
            [line: 7], []},
          nil
        ]},
        with_default: {:integer, [line: 8], []}
      ]
  """

  use TypedStruct.Plugin

  @impl true
  @spec after_definition(keyword()) :: Macro.t()
  def after_definition(_opts) do
    quote do
      def __keys__, do: @ts_fields |> Keyword.keys() |> Enum.reverse()
      def __defaults__, do: Enum.reverse(@ts_fields)
      def __types__, do: Enum.reverse(@ts_types)
    end
  end
end
